daemon/image.go

```diff
@@ -113,12 +121,228 @@ func (daemon *Daemon) GetCachedImage(imgID image.ID, config *containertypes.Conf
+		if err != nil {
+			return "", err
+		}
+		parentImageHistory = makeHistoryWithSource(parentImage)
+	}
+
+	// For each thing we are caching from, see if it matches parentImageHistory
+	type matchStruct struct {
+		cacheFrom   string
+		nextLayerID layer.DiffID
+		createdAt   time.Time
+		history     image.History
+	}
+	matches := make([]matchStruct, 0, len(cache.cacheFromImages))
+	for cacheFromName, cacheFromImageHistory := range cache.cacheFromImageHistories {
+
+		if historiesMatch(cacheFromImageHistory, parentImageHistory) &&
+			// Not only do the histories need to match, but cacheFromImageHistory needs to be long enough
+			// that we can peek at the next piece of "history" (which is in the future for parentImage -
+			// that's why we can cache from it!)
+			(len(cacheFromImageHistory) > len(parentImageHistory)) &&
+			// And the command needs to match too.
+			(strings.Join(cfg.Cmd, " ") == cacheFromImageHistory[len(parentImageHistory)].cmd) {
+
+			match := matchStruct{
+				cacheFrom:   cacheFromName,
+				nextLayerID: cacheFromImageHistory[len(parentImageHistory)].resultingLayerID,
+				createdAt:   cacheFromImageHistory[len(parentImageHistory)].createdAt,
+				history:     cacheFromImageHistory[len(parentImageHistory)].history,
+			}
+			matches = append(matches, match)
+		}
+	}
+
+	// Pluck out the newest layer from the potential matches
+	// Pointer so it can be nil
+	var newestMatch *matchStruct
+	for _, match := range matches {
+		if newestMatch == nil || match.createdAt.After(newestMatch.createdAt) {
+			newestMatch = &match
+		}
+	}
+
+	// If we have a match, build an image from it to represent the next state of the build
+	if newestMatch != nil {
+		// TODO: I just copied this from daemon/commit.go. I'm not sure how/if it should be shared.
+		// I guess the history should just be the real History entry for the next layer?
+		rootFS := image.NewRootFS()
+		var oldHistory []image.History
+		osVersion := ""
+		var osFeatures []string
+		var containerConfig containertypes.Config
+
+		if parentImage != nil {
+			rootFS = parentImage.RootFS
+			oldHistory = parentImage.History
+			osVersion = parentImage.OSVersion
+			osFeatures = parentImage.OSFeatures
+			containerConfig = *parentImage.V1Image.Config
+		}
+
+		if !newestMatch.history.EmptyLayer {
+			rootFS.Append(newestMatch.nextLayerID)
+		}
+
+		newHistory := append(oldHistory, newestMatch.history)
+
+		config, err := json.Marshal(&image.Image{
+			V1Image: image.V1Image{
+				DockerVersion: dockerversion.Version,
+				Config:        cfg,
+				Architecture:  runtime.GOARCH,
+				OS:            runtime.GOOS,
+				// TODO: We didn't really make a container, so I'm just going to leave this emtpy.
+				// Container:       container.ID,
+				// TODO: I think this is supposed to mean, "the contianerConfig we started with", so this makes sense?
+				ContainerConfig: containerConfig,
+				// TODO: What should go in here?
+				Author:  "",
```

Comment:
I've never seen this be anything other than the emtpy string - what should it be?